include/std/experimental/conqueue

```diff
@@ -395,20 +396,39 @@ struct buffer_queue<T, Alloc>::pop_sender {
     buffer_queue& queue;
     Receiver receiver;
 
+    // Constructor of the stop_callback may invoke stop_callback, we use the
+    // following technique to avoid running stop_callback if it is being invoked
+    // from its constructor. we start with count 2, one for the registration
+    // operation, another for cancellation. Once we register stop_callback, we
+    // drop the counter by 1. If stop_callback is invoked, we drop the counter
+    // by 1 again. Thus, if during registration, stop_callback was invoked,
+    // registration code will observe the counter to be 0 and thus understands
+    // it needs to deque the waiter and call stopped on the receiver.
+    atomic<int> stop_count{2};
+
     struct cancel_callback {
       operation& self;
       void operator()() noexcept {
-        auto& cq = *self.sender.queue;
+        auto& cq = self.queue;
+        if (--self.stop_count != 0)
+          // registration code is still running, we are done.
+          return;
+
         unique_lock lock(cq.mutex);
-        // After we acquired lock, the operation might have already completed
-        // and was removed from the queue. Hence, try_remove.
+        // After we acquired the lock, the operation might have already
+        // completed and was removed from the queue. Hence, try_remove.
         if (cq.waiters.try_remove(&self)) {
           lock.unlock();
-          stdexec::set_stopped((Receiver&&)self->receiver);
+          stdexec::set_stopped((Receiver&&)self.receiver);
         }
       }
     };
 
+    using __on_stop = std::optional<typename stdexec::stop_token_of_t<
+        stdexec::env_of_t<Receiver>&>::template callback_type<cancel_callback>>;
+
+    __on_stop stop_callback;
+
     operation(buffer_queue& queue, Receiver&& receiver)
         : queue(queue), receiver(std::move(receiver)) {
       this->complete = [](waiter_base* w) noexcept {
```

Comment:
I think you need to destroy the stop-callback before completing the op here to avoid the stop-callback from executing concurrently and trying to access the buffer_queue after the op has completed.

We already know that the queue item has been dequeued if we're executing the completion callback here, so the cancel_callback isn't going to go and call set_stopped() as the try_remove() will fail.
However, it still might try to access the buffer_queue's mutex concurrently with the op's completion handler executing, which could in theory be about to destroy the buffer_queue object. (although it would need to destroy the buffer_queue object before destroying the operation-state of this op which is probably unlikely for most well-behaved parent operations.